---
title: "fasta2nuccomp: a tool to rapidly summarize the contents of FASTA files containing nucleotides"
author: "Brian J. Knaus and Kelly J. Vining"
date: "`r format(Sys.time(), '%Y, %B %d')`"
output: 
  html_document:
    toc: true
    code_folding: hide
bibliography: fasta2nuccomp_manuscript.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Abstract


Many bioinformatic projects begin with a large FASTA format file, such as a genome.
However, the quality of the contents of these files varies from project to project.
For example, some genomes may include large numbers of ambiguous nucleotides which may present issues to other software used in a project.
This presents a need to rapidly assess the contents of these files.
We present fasta2nuccomp as a tool to query and present summaries of the contents of FASTA files containing nucleotides.
This tool consists of a python script and an RMarkdown document.
The python script can be run on local machines or high performance computing facilities and it rapidly summarizes the nucleotide composition for each sequence in the file.
The RMarkdown provides an example of how this summary can be graphically presented as a report or as publication quality figures.
With fasta2nuccomp we can rapidly assess the contents of FASTA format files so we can move on to downstream procedures with confidence in knowing what the file contains.


Keywords: FASTA file, genome announcement, nucleotide composition, quality control.


## Introduction


A genome is thought to consist of a double stranded polysaccharide chain of nucleotides organized into genes, non-genic sequence, and eventually chromosomes.
Current genome sequencing technology reports this biochemical structure in a text file organized in the FASTA format.
Genomics projects therefore involve managing FASTA files.
Part of this management involves summarizing the contents of these large files in order to validate their contents as well as aid in project organization.
Currently, part of the management of these files involves collection of summaries of the FASTA files from the various workflow outputs.
This creates a situation where a significant amount of project time is invested in determining which summaries are available from which output and how to present this information.


Various tools for summarizing bioinformatic exist, but they are currently distributed among many software programs designed to work on various file formats.
For example, the software FastQC [@andrews2010fastqc; @bioinformatics2011fastqc] summarizes raw sequencing reads in the FASTQ format that are a typical result from Illumina sequencing.
The trimmomatic [@bolger2014trimmomatic] attempts to remove nucleotides from low quality Illumina sequencing data.
In order to infer whether gene space has been well assembled BUSCO [@seppey2019busco] queries for conserved genes from several databases throughout the phylogenetic tree of life.
Blobtools [@laetsch2017blobtools] uses GC content and read coverage to attempt to improve assemblies.
Some of these tools are not specific to FASTA format data, or may be a part of a more computationally expensive analysis that may be undesireable to repeat if only a summary is desired.
A solution specific to data in the FASTA format that provides several summaries of the data, examples of the presentation of these summaries, as well as the flexibility to allow the end user to customize these summaries would aid the management of genomics projects as well as the presentation of this information in genome announcements and pangenomic investigations.


We present fasta2nuccomp, a suite of tools intended to summarize the content of FASTA format files that contain nucleotides.
These tools consist of a python script that summarizes the FASTA file in a tabular format and an RMarkdown script that numerically and graphically summarize the tabular data.
Together, these scripts provide a rapid way to summarize the contents of FASTA files and present this information in publication quality graphics, as well as reports for internal use.
The use of RMarkdown not only provides an example of how to present this data in a visual format, it also provides a working example for how the end user can customize this presentation.


## Methods


The repository fasta2nuccomp contains a python script, an RMarkdown script, example data, and example output.
The python script takes a FASTA format file consisting of nucleotides as input.
The python (â‰¥ 3.8.2) script uses biopython [@cock2009biopython] to read in the file and to collect the length, as well as the (case sensitive) number of nucleotides and IUPAC ambiguous nucleotides.
This information is output to a comma delimited text file.
This script can be run independently or can be called from the RMarkdown script.
The RMarkdown [@rmarkdown] script uses R [@R] and reads in the comma delimited text file and summarizes it.
Graphical summaries are made using ggplot2 [@ggplot2] and ggpubr [@ggpubr] and consist of histograms of A, T, G, and C content, length of sequences, IUPAC ambiguity nucleotides (excluding A, C, G, T, and N), undetermined nucleotides (N), as well as a sanity check for unexpected characters.
Numerical summaries are made as well such as the total number of nucleotides, N50, the number of sequences, and the median for these histograms.
Examples are provided on how to output these graphical summaries to \*.png format and publication quality \*.tiff format.
The *Saccharomyces cerevisiae* [genome](http://sgd-archive.yeastgenome.org/sequence/S288C_reference/genome_releases/) [@engel2014reference] has been included as an example and a form of unit test.
Benchmarking was performed on a MacBook Pro with a 2.6 GHz Quad-Core Intel Core i7 CPU and 16 GB 1600 MHz DDR3 RAM.


## Results

The python script "fasta2nuccomp.py" processed the 12 Mbp yeast genome, consisting of 17 sequences, in 0.5 seconds.
This resulted in a comma delimited text file consisting of 18 rows, a header describing the column contents, and a row for each of the 17 sequences in the S288C genome.
This file included a column for the Id of each sequence, it's length, and the number of each nucleotide (A, C, G, T) as well as each ambiguous nucleotide all in a case sensitive manner (to accommodate masking).
The RMarkdown script ```fasta2nuccomp.Rmd``` reads in this tabular data and demonstrates how to create individual plots for information from the output of the python script and how to make a publication quality composite plot summarizing multiple data features.
An example of this plot is presented in Figure 1.
Execution of this RMarkdown script, including checks for dependencies, the call to the python script, and the rendering of graphics, required 7.5 seconds.


```{r fig, echo=FALSE, fig.align='center', fig.cap="Figure 1. A multipanel summary of the yeast S288C genome. Panels A, B, C, and D summarize the per sequence nucleotide content. Panel E summarizes sequence length as well as presents several numerical summaries of sequence length. Panels F, G, and H summarize the per sequence distribution of IUPAC ambiguous and unexpexted characters. Panel I explores the relationship between G/C content and the length of sequences.", out.width = '100%'}
knitr::include_graphics("S288C_reference_sequence_R64-2-1_20150113.fsa.png")
```


## Discussion

We present fasta2nuccomp as a suite of tools that utilizes the speed of python and Biopython [@cock2009biopython] with the graphics and RMarkdown [@allaire_etal2021] report generation of R [@R] to summarize the contents of FASTA files containing nucleotides.
A python script counts the length and nucleotide content for each sequence and saves this information as a comma delimited text file.
This script can be run on a local workstation or be deployed in a high performance computing environment.
The file resulting from this script can either be moved to a local machine, or processing can simply continue if already working on a local workstation.
These per sequence summaries can be further summarized, such as calculating N50 or the number of sequences in the file, or their distribution can be visualized using graphical tools, such as histograms, in R.
The tool fasta2nuccomp provides an RMarkdown script to demonstrate these tasks and provides the user the option for customization.
The RMarkdown can be compiled into several formats, such as \*.html and \*.pdf, to provide a report that includes descriptive text as well as graphics and tables produced in R.
Because the R code to produce summaries is included in this RMarkdown document the workflow is documented and the end user is also free to customize these examples to fit their specific needs.


Currently, there are a diversity of software tools that output various summaries.
But there does not appear to be a tool to efficiently provides a standardized suite of summaries.
Instead, investigators find themselves collating summaries from several outputs and then presenting this information in a uniform manner.
Some of these summaries are output at the end of a time intensive analysis (such as assembly), a step that is typically not justified if only a summary is desired.
For example, an investigator may obtain an accessioned version of a genome and want to validate the published values.
This is particularly important because some publications may summarize assembled "chromosomes" while accessions versions may include unplaced contigs, the mitochondrion and perhaps a chloroplast.
Having a rapid tool to calculate many summaries provides a quick way to validate if downloaded data files matched the published description.
Because the summary table is saved as a text file and all of the code was made available other users have the potential to invent combinations of this information that the present authors never imagined.
The tool fasta2nuccomp provides rapid summaries of genomes as an initial step for workflows in FASTA based genomic investigations.



# References

