---
title: "nuccomp: rapidly summarize the contents of FASTA and FASTQ files containing nucleotides"
author: "Brian J. Knaus and Kelly J. Vining"
date: "`r format(Sys.time(), '%Y, %B %d')`"
output: 
  html_document:
    toc: true
  word_document:
bibliography: references.bib
csl: oxford-university-press-scimed-author-date.csl
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


---

**To do:**

- Bioinformatics (impact factor: 6.931) - Application Notes 
  - https://academic.oup.com/bioinformatics/pages/instructions_for_authors
  - up to 4 pages; this is approx. 2,600 words or 2,000 words plus one figure 
- export to word?

- Cover letter
- 350 dpi tif figures; single (86 mm) or double (178 mm) column widt
- second version of your figures at low-resolution for use in the review 
- Applications notes: Title page, Short Structured Abstract, Text.
- References: The journal follows Oxford SCIMED style.



---


## Abstract


Bioinformatic projects include the management of large FASTA and FASTQ format files, such as a genomes.
The contents of these files, and their quality, varies from project to project and file to file.
For example, some genomes may include large numbers of ambiguous nucleotides which may present issues to other software used to analyze these files.
This presents a need to rapidly determine the contents of these files.
We present nuccomp as a tool to query and present summaries of the contents of FASTA and FASTQ files containing nucleotides.
This tool consists of a python script and an RMarkdown document.
The python script can be run on local machines or high performance computing facilities and it rapidly summarizes the nucleotide composition for each sequence in each file.
The RMarkdown provides an example of how this summary can be graphically presented as a report or as publication quality figures.
With nuccomp we can rapidly determine the contents of FASTA and FASTQ format files so we can move on to downstream procedures with confidence in knowing the contents of these files.


Keywords: FASTA, FASTQ, genome announcement, nucleotide composition, quality control.


## Introduction


A genome is thought to consist of a double stranded polysaccharide chain of nucleotides organized into genes, intergenic sequence, and eventually chromosomes.
Current genome sequencing technology reports this biochemical structure in a text file organized in the FASTA or FASTQ format [@cock2010sanger].
Genomics projects therefore involve managing FASTA and FASTQ files.
Part of this management involves summarizing the contents of these large files in order to validate their contents as well as aid in project organization.
Currently, part of the management of these files involves collection of summaries of the FASTA files from the various workflow outputs.
This creates a situation where a significant amount of project time is invested in determining which summaries are available from which output and how to present this information.


Various tools for summarizing bioinformatic exist, but they are currently distributed among many software programs designed to work on various file formats.
For example, the software FastQC [@andrews2010fastqc] summarizes raw sequencing reads in the FASTQ format that are a typical result from Illumina sequencing.
The trimmomatic [@bolger2014trimmomatic] attempts to remove nucleotides from low quality Illumina sequencing data.
In order to infer whether gene space has been well assembled BUSCO [@simao2015busco] queries for conserved genes from several databases throughout the phylogenetic tree of life.
Blobtools [@laetsch2017blobtools] uses GC content and read coverage to attempt to improve assemblies.
Some of these tools are not specific to FASTA or FASTQ format data, or may be a part of a more computationally expensive analysis that may be undesireable to repeat if only a summary is desired.
A solution specific to data in the FASTA and FASTQ format that provides several summaries of the data, examples of the presentation of these summaries, as well as the flexibility to allow the end user to customize these summaries would aid the management of genomics projects as well as the presentation of this information in genome announcements and pangenomic investigations.


We present nuccomp, a suite of tools intended to summarize the content of FASTA format files that contain nucleotides.
These tools consist of a python script that summarizes the FASTA file in a tabular format and an RMarkdown script that numerically and graphically summarize the tabular data.
Together, these scripts provide a rapid way to summarize the contents of FASTA files and present this information in publication quality graphics, as well as reports for internal use.
The use of RMarkdown not only provides an example of how to present this data in a visual format, it also provides a working example for how the end user can customize this presentation.


## Methods


The repository nuccomp [@knausb2023] contains two python scripts, an RMarkdown document, example data, and example output.
The python scripts takes a FASTA or FASTQ format file, consisting of nucleotides, as input.
The output from the python scripts are summarized graphically and with tables using the RMarkdown document.
The result can be an html document suitable for internal reports as well as publication quality graphics (\*.png, *.tiff, and other formats).


The python script nuccomp.py takes as input a FASTA or FASQ format file, which may or may not be compressed with gzip, as input.
The python (≥ 3.8.2) script uses biopython (≥ 1.79; @cock2009biopython) to read in the file and to collect the length, as well as the (case sensitive) number of nucleotides and IUPAC ambiguous nucleotides.
This information is output to a comma delimited text file.
This script may be run independently or may be called from the RMarkdown document.


The python script motif_counter.py divides each sequence in a FASTA or FASTQ file into windows of user specified length and then counts the number of motifs found with a regular expression within each window.
The python (≥ 3.8.2) script uses biopython (≥ 1.79; @cock2009biopython) to read in the file and custom python to divide each sequence into windows and count the number of times a regular expression is matched within each window.
This count is converted to a rate by dividing each count by it's window size, scaled by subtracting the minimum and dividing by the maximum (after subtracting the minimum).
Creating a rate (motifs per position) helps to accommodate the last window of each sequence which is expected to be smaller than the user specified window size.
This scaled rate can then be inverted by subtracting it from one and is reported in the BED format output as the 'score' column 5.
The coordinates for each window, the count, and score are reported in a BED format file.


The RMarkdown [@rmarkdown] document uses R [@R] and reads in the comma delimited text file resulting from nuccomp.py and the BED file from motif_counter.py and summarizes them.
Graphical summaries are made using ggplot2 [@ggplot2] and ggpubr [@ggpubr] and consist of histograms of A, T, G, and C content, length of sequences, IUPAC ambiguity nucleotides (excluding A, C, G, T, and N), undetermined nucleotides (N), as well as a sanity check for unexpected characters.
Numerical summaries are made as well such as the total number of nucleotides, N50, the number of sequences, and the median for these histograms.
Examples are provided on how to output these graphical summaries to \*.png format and publication quality \*.tiff format.


The *Saccharomyces cerevisiae* genome (http://sgd-archive.yeastgenome.org/sequence/S288C_reference/genome_releases/; @engel2014reference) has been included as an example and a form of unit test.
In addition, the potato (*Solanum tuberosum*; GCA_001642375.1), hemp (*Cannabis sativa*; GCF_900626175.2), mint (*Mentha longifolia*; GCA_001642375.2), rice (*Oryza sativa*; GCF_001433935.1) cottonwood (*Populus trichocarpa*; GCF_000002775.5), hazelnut (*Corylus avellana*; GCA_901000735.2), and *Arabidopsis thaliana* (GCF_000001735.4) genomes were used for benchmarking.
Benchmarking was performed on computer using an Intel(R) Core(TM) i7-2600K CPU @ 3.40GHz CPU running Ubuntu 2022.04.2 LTS and R 4.3.0.
The repository nuccomp is currently available on GitHub [@knausb2023].


## Results

The python script "nuccomp.py" processed the 12 Mbp yeast genome, consisting of 17 sequences, in less than 1 second (large genomes consisting of more nucleotides and/or sequences should require more time).
This resulted in a comma delimited text file consisting of 18 rows, a header describing the column contents, and a row for each of the 17 sequences in the S288C genome.
This file included a column for the 'ID' of each sequence, it's length, and the number of each nucleotide (A, C, G, T) as well as each ambiguous nucleotide all in a case sensitive manner (to accommodate masking).
The python script "motif_counter.py" summarized the yeast genome into windows into counts of the motif 'CG' and reported this to a BED format file.
The RMarkdown script ```nuccomp.Rmd``` read in this these data and demonstrates how to create individual plots for information from the output of the python script and how to make a publication quality composite plot summarizing multiple data features.
An example of this plot is presented in Figure 1.
Execution of this RMarkdown script, including checks for dependencies, calls to the python scripts, and the rendering of graphics, required less than 10 seconds.


```{r fig, echo=FALSE, fig.align='center', fig.cap="**Figure 1.** Summary of the yeast S288C assembly (R64-2-1_20150113). Panel A is the distribution of sequence lengths. Panel B presents the distribution for per sequence, nucleotide frequencies. Panel C is an ideogram presenting the lengths (height) of sequences. Each sequence is partitioned into 15 kbp windows where the width and color was determined by the inverse scaled count of the motif 'CG' within each window. Panels D, E, and F summarize the occurrence of IUPAC ambiguous nucleotides (wWsSmMkKrRyY), any nucelotide (nN), and any unexpected nucleotides on a per sequence basis. Panel G presents the relationship between the length of each sequence and it’s G/C content.", out.width = '70%'}
knitr::include_graphics("../S288C_reference_sequence_R64-2-1_20150113.png")
```


In order to demonstrate the convenience of nuccomp, it was implemented on several publicly available genomes.
The potato (*Solanum tuberosum*) genome, consisted of 705 Mbp and an N50 of 1.3 Mbp, required 46.6 seconds.
The hemp (*Cannabis sativa*) genome, consisted of 876 Mbp and an N50 of 91 Mbp, required	45.7 seconds.
The mint (*Mentha longifolia*) genome, consisted of 468 Mbp and an N50 of 37 Mbp, required 29.0 seconds.
The rice (*Oryza sativa*) genome, consisted of 374 Mbp and an N50 of 29 Mbp, required 25.1 seconds.
The cottonwood (*Populus trichocarpa*) genome, consisted of 392 Mbp and an N50 of 19 Mbp, required 24.6 seconds.
The hazelnut (*Corylus avellana*) genome, consisted of 369 Mbp and an N50 of 36 Mbp, required 23.7 seconds.
The *Arabidopsis thaliana* genome, consisted of 119 Mbp and an N50 of 23 Mbp required 12.4 seconds.
Processing time included the running of python code and compiling ('kniting') the RMarkdown, includeing text input and graphics rendering.
Benchmarking results can be found at the GitHub repository (@knausb2023) in the directory 'benchmarks'.


## Discussion


The repository nuccomp consists of a suite of tools that utilizes the speed of python and Biopython [@cock2009biopython] with the graphics and statistical resources of R [@R] combined with RMarkdown [@allaire_etal2021] for report generation to summarize the contents of FASTA and FASTQ files containing nucleotides.
Genomics project typically include aggregating files of sequence data, such as genomes acquired from NCBI or sequencing consortia, as well as sequencing libraries that may be assembled multiple times to explore assembly options.
This creates a need to rapidly inventory the contents of these FASTA and FASTQ format sequence files.
The repository nuccomp provides a rapid solution to help inventory the contents of these files and presents this information in graphical and tabular forms.


Other tools


Several currently available software tools contain functions to output various summaries of FASTA and FASTQ files.


QUAST



FastQC

BlobTools


 BUSCO

, such as QUAST [@gurevich2013quast] and SeqKit (refs).
But there does not appear to be a tool to efficiently provides a standardized suite of summaries.
Instead, investigators find themselves collating summaries from several outputs and then presenting this information in a uniform manner.
Some of these summaries are output at the end of a time intensive analysis (such as assembly), a step that is typically not justified if only a summary is desired.
For example, an investigator may obtain an accessioned version of a genome and want to validate the published values.
This is particularly important because some publications may summarize assembled "chromosomes" while accessions versions may include unplaced contigs, plus non-autosomal contigs representing the mitochondrion and, for plants, the chloroplast.




Literate programming


The use of an RMarkdown document facilitates the summary of the data and also provides a high level code example for the summaries made.
This RMarkdown document is a text file that is compiled, or 'knit', into several formats, such as \*html or \*pdf.
The text file includes text as well as code chunks, as an attempt at literate programming.
This means that the N50 summary statistic in the report includes the code used to generate the summary, providing transparency and educational value.
Similarly, the graphical summaries include code examples for how the plots were generated.
It is hoped that by providing this functionality will allow users to create their own innovations derived from this foundation.


Having a rapid tool to calculate many summaries provides a quick way to validate if downloaded data files matched the published description.
Because the summary table is saved as a text file and all of the code was made available other users have the potential to invent combinations of this information that the present authors never imagined.
The tool nuccomp provides rapid summaries of genomes as an initial step for workflows in FASTA based genomic investigations.


The study of genomics could currently be characturized as the study of text files in the FASTA and FASTQ formats.
These files are acquired through data centers and sequencing consortia as benchmarks of quality and to build upon our existing knowledge base.
These files are also  generated at multiple steps during research projects, such as receiving sequence libraries from a sequencing center, the result of assembling these sequences, as well as any results from attempts at scaffolding these assemblies into more contiguous representations of chromosomes and genomes.
The repository nuccomp contributes to the collection of similar tools designed to summarize the contents of FASTA and/or FASTQ data (@andrews2010fastqc, @gurevich2013quast, @laetsch2017blobtools).
Through providing nuccomp as a high level solution, using RMarkdown, it is hoped that this will empower investigators to not only prepare reports and summaries as demonstrated here, but will also provide a foundation that inspires users to innovate their own representations of genomic information.


## Acknowledgements

This work was supported by OregonCBD (Independence, Oregon, USA) through a gift to the Oregon State University Foundation.


## References


