---
title: "Benchmarking summary"
author: "Insert authors here"
date: "`r format(Sys.time(), '%Y, %B %d')`"
output:
  html_document:
    toc: true
    code_folding: hide
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.align = "center")
knitr::opts_chunk$set(fig.height = 4)
knitr::opts_chunk$set(fig.width = 4)
st1 <- Sys.time()
```



```{r, include=FALSE}
bnchmrk <- data.frame( 
  Mentha <- structure(c(468946732, 307, 36.8542295372372, 37537474, 46699537, 
11198, 4997, 21.1178591251373), dim = c(8L, 1L), dimnames = list(
    c("Nucleotides", "Sequences", "GC", "N50", "Maximum", "Median", 
    "Minimum", "Seconds"), NULL)),
  Oryza <- structure(c(995045545, 1954, 45.0162325986797, 1922674, 24906701, 
141848.5, 1324, 33.5186097621918), dim = c(8L, 1L), dimnames = list(
    c("Nucleotides", "Sequences", "GC", "N50", "Maximum", "Median", 
    "Minimum", "Seconds"), NULL)),
  Corylus <- structure(c(369779143, 11, 35.8991454083842, 36653616, 50950907, 
30273198, 22421508, 16.9769184589386), dim = c(8L, 1L), dimnames = list(
    c("Nucleotides", "Sequences", "GC", "N50", "Maximum", "Median", 
    "Minimum", "Seconds"), NULL)),
  Arabidopsis <- structure(c(119668634, 7, 36.0559867100791, 23459830, 30427671, 
19698289, 154478, 10.8701975345612), dim = c(8L, 1L), dimnames = list(
    c("Nucleotides", "Sequences", "GC", "N50", "Maximum", "Median", 
    "Minimum", "Seconds"), NULL)),
  Populus <- structure(c(434289848, 1447, 33.7475523113723, 19465461, 50495391, 
6403, 657, 21.7682278156281), dim = c(8L, 1L), dimnames = list(
    c("Nucleotides", "Sequences", "GC", "N50", "Maximum", "Median", 
    "Minimum", "Seconds"), NULL))
)
names(bnchmrk) <- c("Mentha", "Oryza", "Corylus", "Arabidopsis", "Populus")
```


```{r, include=FALSE}
bnchmrk <- bnchmrk[, sort.int(unlist(bnchmrk['Seconds', ]), index.return = TRUE, decreasing = TRUE)$ix]
knitr::kable(bnchmrk, format.args = list( digits = 2, nsmall = 3, scientific = F, big.mark = ","))
knitr::kable(t(bnchmrk), format.args = list( digits = 3, nsmall = 1, scientific = F, big.mark = "," ))
```




```{r}
# unlink("*_nuccomp.csv")
# unlink("benchmarking.csv")
```

```{r}
# system("Rscript -e \"rmarkdown::render(\'report_arabidopsis.Rmd\')\"")
# system("Rscript -e \"rmarkdown::render(\'report_corylus.Rmd\')\"")
# system("Rscript -e \"rmarkdown::render(\'report_mentha.Rmd\')\"")
# system("Rscript -e \"rmarkdown::render(\'report_oryza.Rmd\')\"")
# system("Rscript -e \"rmarkdown::render(\'report_populus.Rmd\')\"")
```


## Genome Inventory


In order to test the perfomance of our software we processed several plant genomes.
The files for each accession, as well as additional information about each assembly, can be found at NCBI (https://www.ncbi.nlm.nih.gov/).


```{r}
genomes <- matrix(c("Arabidopsis", "GCF_000001735.4", "GCF_000001735.4_TAIR10.1_genomic.fna.gz",
"Corylus", "GCA_901000735.2", "GCA_901000735.2_CavTom2PMs-1.0_genomic.fna.gz",
"Mentha", "GCA_001642375.2", "GCA_001642375.2_Mlong_CMEN585_v3_genomic.fna.gz",
"Oryza", "GCA_019925245.1", "GCA_019925245.1_MQ-UA-ANU_Oaus-KR_1.0_genomic.fna.gz",
"Populus", "GCF_000002775.4", "GCF_000002775.4_Pop_tri_v3/GCF_000002775.4_Pop_tri_v3_genomic.fna.gz"), 
ncol = 3, byrow = TRUE)
colnames(genomes) <- c("Taxon", "Accession", "File name")

knitr::kable(genomes, caption = "**Table 1.** Accessions and common names for genomes used for benchmarking.")
```


Automated downloading of these genomes can be perfomed using `wget` by uncommenting the below commands.
(And recommenting them when you're done so you avoid downloading them again.)


```{bash}
# wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/735/GCF_000001735.4_TAIR10.1/GCF_000001735.4_TAIR10.1_genomic.fna.gz
# gunzip GCF_000001735.4_TAIR10.1_genomic.fna.gz
# wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/901/000/735/GCA_901000735.2_CavTom2PMs-1.0/GCA_901000735.2_CavTom2PMs-1.0_genomic.fna.gz
# gunzip GCA_901000735.2_CavTom2PMs-1.0_genomic.fna.gz
# wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/642/375/GCA_001642375.2_Mlong_CMEN585_v3/GCA_001642375.2_Mlong_CMEN585_v3_genomic.fna.gz
# gunzip GCA_001642375.2_Mlong_CMEN585_v3_genomic.fna.gz
# wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/019/925/245/GCA_019925245.1_MQ-UA-ANU_Oaus-KR_1.0/GCA_019925245.1_MQ-UA-ANU_Oaus-KR_1.0_genomic.fna.gz
# gunzip GCA_019925245.1_MQ-UA-ANU_Oaus-KR_1.0_genomic.fna.gz
# wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/775/GCF_000002775.4_Pop_tri_v3/GCF_000002775.4_Pop_tri_v3_genomic.fna.gz
# gunzip GCF_000002775.4_Pop_tri_v3/GCF_000002775.4_Pop_tri_v3_genomic.fna.gz
```


## Benchmarking


The following lines of code render an RMarkdown report that calls `fasta2nuccomp.py` and collects the execution time.
You will be expected to manage your python/conda environment and make sure biopython is installed in the environment that is called from RMarkdown.
A potential solution is to use the R package `reticulate` (https://CRAN.R-project.org/package=reticulate) to help the interface between R and python.
Note that the `rm` call will delete the benchmarking file.
This is because the rendering of the RMarkdown will append to this file, meaning you have to delete this file to begin the collection of more information.
Once the benchmarking has been run, and the `benchmarking.csv` file has been created, these lines may be commented out to avoid benchmarking and focus on downstream processing of the benchmarking data.


```{bash, results = 'hide'}
# conda activate biopython
# rm benchmarking.csv
# rm *_nuccomp.csv
# Rscript -e "rmarkdown::render('report_arabidopsis.Rmd')"
# Rscript -e "rmarkdown::render('report_corylus.Rmd')"
# Rscript -e "rmarkdown::render('report_mentha.Rmd')"
# Rscript -e "rmarkdown::render('report_oryza.Rmd')"
# Rscript -e "rmarkdown::render('report_populus.Rmd')"
```


## Summarization


Summarization of each assembly and execution time are presented in Table 2.
The evaluated assemblies ranged from 100 million to almost 1 billion nucleotides arranged in a total number of sequences ranging from seven to almost two thousand provided assemblies with a diversity of characteristics.
The execution time includes the execution of `fasta2nuccomp.py` and the rendering of an RMarkdown document, including a graphical summary, for each genome.
The execution times were under one minute for all evaluated genomes.


```{r}
bnchmrk <- read.table("benchmarking.csv", header = FALSE, sep = ",")

colnames(bnchmrk) <- c("Taxon", "Nucleotides", "Sequences", "GC", "N50", "Maximum", "Median", "Minimum", "Seconds")

# bnchmrk <- bnchmrk[sort.int(bnchmrk$Nucleotides, index.return = TRUE, decreasing = TRUE)$ix, ]
bnchmrk <- bnchmrk[sort.int(bnchmrk$Sequences, index.return = TRUE, decreasing = TRUE)$ix, ]

# knitr::kable(bnchmrk, 
#              format.args = list( digits = 2, nsmall = 3, scientific = F, big.mark = ","),
#              row.names = FALSE)

#bnchmrk$GC <- as.numeric(bnchmrk$GC)

knitr::kable(bnchmrk, 
             format.args = list( digits = 3, 
                                 nsmall = 1, 
                                 scientific = F, big.mark = ","),
             row.names = FALSE, caption = "**Table 2.** Genome summaries produced by fasta2nuccomp.py and execution time. ")

```



