---
title: "fasta2nuccomp"
author: "Insert authors here"
date: "`r format(Sys.time(), '%Y, %B %d')`"
output:
  html_document:
    toc: true
    code_folding: hide
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.align = "center")
knitr::opts_chunk$set(fig.height = 4)
knitr::opts_chunk$set(fig.width = 4)
st1 <- Sys.time()
```



## FASTA file input

Edit code here to specify your FASTA file.
You'll also want to include fasta2nuccomp.py and it's full path.


```{r}
st2 <- Sys.time()
#my_infile <- "S288C_reference_sequence_R64-2-1_20150113.fsa"
#my_infile <- "GCA_001642375.2_Mlong_CMEN585_v3_genomic.fna"
#my_infile <- "GCA_901000735.2_CavTom2PMs-1.0_genomic.fna"
my_infile <- "GCF_000001735.4_TAIR10.1_genomic.fna"
my_out <- basename(my_infile)
# Only run python script if it's outfile does not exist. 
if( !file.exists( paste(my_out, "_nuccomp.csv", sep = "") ) ){
  system(paste("~/gits/fasta2nuccomp/fasta2nuccomp.py", my_infile))
}
```


```{r}
st3 <- Sys.time()
st3-st2
```


## Post-processing


Once we have summarized our FASTA data we'll want to present it.
This can be accomplished with summary statistics and graphics.
Examples are below.


```{r}
my_nucs <- readr::read_csv(paste(my_out, "_nuccomp.csv", sep = ""),
                           show_col_types = FALSE)
#my_nucs
```



```{r}
my_lens <- sort(my_nucs$Length, decreasing = TRUE)
my_mid <- sum(my_lens)/2
n50 <- my_lens[cumsum(my_lens) >= my_mid][1]
#format(n50, big.mark = ",")
```



## Individual panels


Graphical summaries can be created in R as follows.
These individual panels can be combined into one 'dashboard' below.


```{r}
library(ggplot2)
```


```{r}
my_df <- data.frame(Length = my_nucs$Length,
                    GC = rowSums(my_nucs[,c("c", "C", "g", "G")])
                    )

my_max <- max(my_df$Length)
my_med <- median(my_df$Length)
my_min <- min(my_df$Length)
my_seqs <- length(my_df$Length)
my_nzero <- length(my_df$Length > 0)
my_nna <- length(!is.na(my_df$Length))
my_len <- sum(my_df$Length)

#my_df$GC <- my_df$GC/my_df$Length
#
my_denom <- 1e3
my_df$Length <- my_df$Length/my_denom


my_text <- paste("Nucleotides:\n", format(my_len, big.mark = ","),
                 "\nN50: ", format(n50, big.mark = ","),
                 "\nN seqs: ", format(my_seqs, big.mark = ","),
                 "\nMax: ", format(my_max, big.mark = ","),
                 "\nMedian: ", format(my_med, big.mark = ","),
                 "\nMin: ", format(my_min, big.mark = ","),
                 sep = ""
                 )
```


```{r}

pl <- ggplot(my_df, aes(x = Length))
#pl <- pl + geom_histogram(binwidth=2e4, fill = "blue")

pl <- pl + geom_histogram(binwidth = 2e1, fill = "blue")
pl <- pl + theme_bw()
pl <- pl + geom_vline(xintercept = n50/my_denom, color = "#B22222", size = 1.5,
                      linetype="dashed")
my_ymax <- max(ggplot_build(pl)$data[[1]]$y)/2
#
pl <- pl + annotate("text", x = my_max/my_denom * 0.95, y = my_ymax,
                    label = my_text, adj = 1)

pl <- pl + xlab("Length (Kbp)")
pl <- pl + ylab("Count")

#pl
```


```{r}
my_df <- as.data.frame(rowSums(my_nucs[,c("a","A")])/my_nucs$Length)
names(my_df) <- "nucs"
my_med <- median(my_df$nucs)

pua <- sum(my_nucs[,c("A")])/sum(rowSums(my_nucs[,c("a","A")]))

my_text <- paste("Median: ", round(my_med, 3),
                 "\nUppercase: ", round(pua * 100, 1),
                 "%",
                 sep = ""
                 )


pa <- ggplot(my_df, aes(x = nucs))
pa <- pa + geom_histogram(binwidth=0.01, fill = "#5157FB")
pa <- pa + theme_bw()
pa <- pa + xlim(c(0,1))
pa <- pa + xlab("Adenine (A, a)")
#pa <- pa + annotate("text", x = 0.95, y=1.5, label= my_text, adj = 1)
my_ymax <- max(ggplot_build(pa)$data[[1]]$y)/2
pa <- pa + annotate("text", x = 0.95, y=my_ymax, label= my_text, adj = 1)
pa <- pa + ylab("Count")

#pa
```


```{r}
my_df <- as.data.frame(rowSums(my_nucs[,c("c","C")])/my_nucs$Length)
names(my_df) <- "nucs"
my_med <- median(my_df$nucs)

puc <- sum(my_nucs[,c("C")])/sum(rowSums(my_nucs[,c("c","C")]))

my_text <- paste("Median: ", round(my_med, 3),
                 "\nUppercase: ", round(puc * 100, 1),
                 "%",
                 sep = ""
                 )

pc <- ggplot(my_df, aes(x = nucs))
pc <- pc + geom_histogram(binwidth=0.01, fill = "#DD0A17")
pc <- pc + theme_bw()
pc <- pc + xlim(c(0,1))
pc <- pc + xlab("Cytosine (C, c)")
#pc <- pc + annotate("text", x = 0.95, y=1.5, label= my_text, adj = 1)
my_ymax <- max(ggplot_build(pc)$data[[1]]$y)/2
pc <- pc + annotate("text", x = 0.95, y=my_ymax, label= my_text, adj = 1)
pc <- pc + ylab("Count")

#pc
```


```{r}
my_df <- as.data.frame(rowSums(my_nucs[,c("g","G")])/my_nucs$Length)
names(my_df) <- "nucs"
my_med <- median(my_df$nucs)

pug <- sum(my_nucs[,c("G")])/sum(rowSums(my_nucs[,c("g","G")]))

my_text <- paste("Median: ", round(my_med, 3),
                 "\nUppercase: ", round(pug * 100, 1),
                 "%",
                 sep = ""
                 )

pg <- ggplot(my_df, aes(x = nucs))
pg <- pg + geom_histogram(binwidth=0.01, fill = "#1CBE20")
pg <- pg + theme_bw()
pg <- pg + xlim(c(0,1))
pg <- pg + xlab("Guanine (G, g)")
#pg <- pg + annotate("text", x = 0.95, y=1.5, label= my_text, adj = 1)
my_ymax <- max(ggplot_build(pg)$data[[1]]$y)/2
pg <- pg + annotate("text", x = 0.95, y=my_ymax, label= my_text, adj = 1)
pg <- pg + ylab("Count")

#pg
```


```{r}
my_df <- as.data.frame(rowSums(my_nucs[,c("t","T")])/my_nucs$Length)
names(my_df) <- "nucs"
my_med <- median(my_df$nucs)

put <- sum(my_nucs[,c("T")])/sum(rowSums(my_nucs[,c("t","T")]))

my_text <- paste("Median: ", round(my_med, 3),
                 "\nUppercase: ", round(put * 100, 1),
                 "%",
                 sep = ""
                 )

pt <- ggplot(my_df, aes(x = nucs))
pt <- pt + geom_histogram(binwidth=0.01, fill = "#E6E431")
pt <- pt + theme_bw()
pt <- pt + xlim(c(0,1))
pt <- pt + xlab("Thymine (T, t)")
#pt <- pt + annotate("text", x = 0.95, y=1.5, label= my_text, adj = 1)
my_ymax <- max(ggplot_build(pt)$data[[1]]$y)/2
pt <- pt + annotate("text", x = 0.95, y=my_ymax, label= my_text, adj = 1)
pt <- pt + ylab("Count")

#pt
```


```{r}
my_df <- as.data.frame(rowSums(my_nucs[,c("w","W", "s", "S", "m", "M", "k", "K", "r", "R", "y", "Y")])/my_nucs$Length)
names(my_df) <- "nucs"

pu <- ggplot(my_df, aes(x = nucs))
pu <- pu + geom_histogram(binwidth=0.01, fill = "#AD26FA")
pu <- pu + theme_bw()
pu <- pu + xlim(c(0,1))
# pu <- pu + xlab("IUPAC ambiguous (w,W, s, S, m, M, k, K, r, R, y, Y)")
pu <- pu + xlab("IUPAC ambiguous")
pu <- pu + ylab("Count")
pu <- pu + theme(axis.title.x = element_text(size = 10, hjust = 0.5),
                 axis.text.x = element_text(angle =  0, hjust = 0.5, size=8)
                 )
pu <- pu + theme(axis.title.y = element_text(size = 10, hjust = 0.5))

#pu
```


```{r}
my_df <- as.data.frame(rowSums(my_nucs[,c("n","N")])/my_nucs$Length)
names(my_df) <- "nucs"

pn <- ggplot(my_df, aes(x = nucs))
pn <- pn + geom_histogram(binwidth=0.01, fill = "#AD26FA")
pn <- pn + theme_bw()
pn <- pn + xlim(c(0,1))
pn <- pn + xlab("Any base (N,n)")
pn <- pn + ylab("Count")
pn <- pn + theme(axis.title.x = element_text(size = 10, hjust = 0.5),
                 axis.text.x = element_text(angle =  0, hjust = 0.5, size=8)
                 )
pn <- pn + theme(axis.title.y = element_text(size = 10, hjust = 0.5))

#pn
```


```{r}
my_df <- as.data.frame(
  my_nucs$Length - 
  rowSums(my_nucs[,c("a", "A", "c", "C", "g", "G", "t", "T", "w", "W", "s", "S", "m", "M", "k", "K", "r", "R", "y", "Y", "n", "N")])/
  my_nucs$Length)
names(my_df) <- "nucs"

pe <- ggplot(my_df, aes(x = nucs))
pe <- pe + geom_histogram(binwidth=0.01, fill = "#AD26FA")
pe <- pe + theme_bw()
pe <- pe + xlim(c(0,1))
#pe <- pe + xlab("Unexpected characters")
pe <- pe + xlab("Unexpected")
pe <- pe + ylab("Count")
pe <- pe + theme(axis.title.x = element_text(size = 10, hjust = 0.5),
                 axis.text.x = element_text(angle =  0, hjust = 0.5, size=8)
                 )
pe <- pe + theme(axis.title.y = element_text(size = 10, hjust = 0.5))

#pe
```


```{r}
my_df <- data.frame(Length = my_nucs$Length,
                    GC = rowSums(my_nucs[,c("c", "C", "g", "G")])
                    )
my_df$GC <- my_df$GC/my_df$Length
my_df$Length <- my_df$Length/1e6

pgc <- ggplot(my_df, aes(x = GC, y = Length))
pgc <- pgc + geom_point(size = 4, shape = 21, fill = "#B2222299")
pgc <- pgc + theme_bw()
pgc <- pgc + xlim(c(0,1))
# pgc <- pgc + ylab("Length (Kbp)")
pgc <- pgc + ylab("Mbp")
pgc <- pgc + theme(axis.title.x = element_text(size = 10, hjust = 0.5),
                   axis.text.x = element_text(angle =  0, hjust = 0.5, size=8)
                   )
pgc <- pgc + theme(axis.title.y = element_text(size = 10, hjust = 0.5))

#pgc

```


## Dashboard


For internal purposes we may want a detailed report, as presented above.
If we want to publish this information we may want to summarize this information in a single, multipanel graphic, or dashboard perspective.
This is illustrated below.


```{r, fig.height=8, fig.width=8, fig.cap="**Figure X.** Summary of the genome. Panels **A**, **B**, **C**, and **D** summarize the occurrence of each nucleotide on a per sequence/contig basis. Lowercase characters may indicate that some form of 'masking' may have been implemented. Panel **E** summarizes the length of each sequence/contig, a dashed vertical red line indicates the N50 for the set of sequences/contigs. Panel **F** summarizes the number of positions where [IUPAC ambiguity codes](https://en.wikipedia.org/wiki/Nucleic_acid_sequence#Notation) occured (excluding A, a, C, c, G, g, T, t, N, and n). Panel **G** summarizes the occurence of 'any nucleotide' (N, n) in each sequence/contig. Panel **H** summarizes the occurence of unexpected characters in the sequences, and is expected to be zero. Panel **I** presents the relationship between the length of each sequence/contig and it's G/C content."}
library(ggpubr)

ggarrange(
  ggarrange(pa, pc, pt, pg, ncol = 2, nrow = 2, labels = c("A", "B", "C", "D")),
  pl,
  ggarrange(pu, pn, pe, pgc, ncol = 4, nrow = 1, 
            labels = c("F", "G", "H", "I"),  
            font.label = list(size = 10, color = "black", face = "bold", family = NULL)),
  ncol = 1, nrow = 3,
  labels = c("", "E", ""),
  heights = c(2.5, 2, 1))


# ggsave(filename = paste(my_out, ".png", sep = ""), device = "png",
#        width = 6.5, height = 6.5, units = "in")
# ggsave(filename = paste(my_out, ".tiff", sep = ""), device = "tiff",
#        width = 6.5, height = 6.5, units = "in", compression = "lzw")
```


For the purpose of publication we may desire to save this "dashboard" as a graphic.
After the above code has been run, the ```ggsave()``` function can be uncommented and run.
You'll want to include a meaningful filename.
This will save the dashboard as a ```*.png``` or ```*.tiff``` graphic.


We may also want a tabular presentation of this summary.
Below is an example of tabular output.
This output may be presented as it is below, or it may be combined into a larger table, for example, one including multiple genomes.


```{r, comment=""}
my_gc <- sum( my_nucs[, c("c", "C", "g", "G")] )/sum(my_nucs[, c("a", "A", "c", "C", "g", "G", "t", "T")])
my_med <- median(my_nucs$Length)

st4 <- Sys.time()
my_st <- st4 - st1


my_summary <- rbind(my_len, my_seqs, my_gc * 100, n50, my_max, my_med, my_min, my_st)
#row.names(my_summary) <- c("Nucleotides", "Sequences", "GC", "N50", "Maximum", "Median", "Minimum")
row.names(my_summary) <- c("Nucleotides", "Sequences", "GC", "N50", "Maximum", "Median", "Minimum", "Seconds")
colnames(my_summary) <- "Arabidopsis"

knitr::kable(my_summary, format.args = list( digits = 2, nsmall = 3, scientific = F))

write.table(t(my_summary), file = "benchmarking.csv",
            append = TRUE, sep = ",",
            row.names = TRUE, col.names = FALSE)

# dput( my_summary )


```




## Session information


After performing the above operations we may wish to benchmark the execution time in order to determine how feasible it is to repeat this analysis.
This has been facilitated above by taking time stamps with the function ```Sys.time()```.
We can now use these timestamps to evaluate total execution time.


```{r}
st5 <- Sys.time()
st5 - st1
```


Lastly, we can use the ```sessionInfo()``` function to report the version of ```R()``` we are using as well as the packages and their versions.
This is important information to include if we want to report an issue.


```{r}
sessionInfo()
```

